sklearn/feature_selection/_base.py

```diff
@@ -78,15 +79,22 @@ def transform(self, X):
         X_r : array of shape [n_samples, n_selected_features]
             The input samples with only the selected features.
         """
-        # note: we use _safe_tags instead of _get_tags because this is a
-        # public Mixin.
-        X = self._validate_data(
-            X,
-            dtype=None,
-            accept_sparse="csr",
-            force_all_finite=not _safe_tags(self, key="allow_nan"),
-            reset=False,
-        )
+        output_config_dense = _get_output_config("transform", estimator=self)["dense"]
+        if hasattr(X, "iloc") and output_config_dense == "pandas":
+            # Only perform input validation on feature names and n_features when the
+            # output is a dataframe, preserving the dataframe for _transform to mask
+            self._check_feature_names(X, reset=False)
+            self._check_n_features(X, reset=False)
+        else:
+            # note: we use _safe_tags instead of _get_tags because this is a
+            # public Mixin.
+            X = self._validate_data(
+                X,
+                dtype=None,
+                accept_sparse="csr",
+                force_all_finite=not _safe_tags(self, key="allow_nan"),
+                reset=False,
+            )
```

Comment:
thinking of when I go and teach people how to write their own estimator, this makes it quite hard to explain what they should do if their estimator selects features and could keep pandas dtypes.

I don't shy away from telling people to use private API, but this is kinda more than one could remember to teach/learn and use.

Reply:
If it is only for feature selection, they can inherit from `SelectorMixin` and define a `_get_support_mask`, and it should "just work".

For other transformers, I would teach "if you want pandas output, then return pandas all the time with the dtypes preserved". The `transform="default"` gives the estimator the option of returning a dataframe as long as it is an "array-like". Our code is more complex because we want to return ndarrays for `transform="default"`.

Reply:
hmm. Still very skeptical. But I'm willing to see this merged and hopefully develop our related private utils for them to be third party usable.

Reply:
From a previous discussion/comment: it seems like you don't need to add the lines that refer to private methods to make the "preserve dtype" part work. It makes sense to add them, but they aren't strictly required. You do need to know that you should not call `_validate_data` because it will result in a conversion of DF to ndarray.

Not sure this is any easier to teach. Might even make it harder :-/